vmm/src/vcpu_throttling.rs

```diff
@@ -0,0 +1,412 @@
+// Copyright © 2025 Cyberus Technology GmbH
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use std::ops::Deref;
+use std::sync::{Arc, Condvar, Mutex, MutexGuard};
+use std::thread;
+use std::thread::{sleep, JoinHandle};
+use std::time::{Duration, Instant};
+
+use vm_migration::Pausable;
+
+use crate::cpu::CpuManager;
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+enum ThreadState {
+    /// Waiting for next event.
+    Waiting,
+    /// Ongoing vCPU throttling.
+    ///
+    /// The inner value shows the current throttling percentage in range `1..=99`.
+    Throttling(u8 /* `1..=99` */),
+    /// Thread is shutting down gracefully.
+    Exiting,
+}
+
+impl ThreadState {
+    /// Returns the currently throttling level in percent.
+    fn throttle_percent(&self) -> u8 {
+        match self {
+            ThreadState::Throttling(val) => *val,
+            _ => 0,
+        }
+    }
+}
+
+/// Type to indicate state change acknowledgments between the handler and the
+/// thread.
+///
+/// See:
+/// - [`VcpuThrottleThreadHandle::CHANGE_ACK`]
+/// - [`VcpuThrottleThreadHandle::CHANGE_NACK`]
+type StateChangeAckT = bool;
+
+type SynchronizedStateInner = (StateChangeAckT, ThreadState);
+type SynchronizedState = (Mutex<SynchronizedStateInner>, Condvar);
+
+/// Handler for controlling the vCPU throttle thread.
+///
+/// vCPU throttling is needed for live-migration of memory-intensive workloads.
+/// The current design assumes that all vCPUs are throttled equally.
+///
+/// # States and Transitions
+/// - `Waiting` -> `Throttle(x %)`, `Exit`
+/// - `Throttle(x %)` -> `Waiting`, `Throttle(y %)`
+/// - `Exit`
+pub struct VcpuThrottleThreadHandle {
+    /// Thread state wrapped by synchronization primitives.
+    shared_state: Arc<SynchronizedState>,
+    /// The underlying thread handle.
+    // Option so that we can .take() the value later
+    thread_handle: Option<JoinHandle<()>>,
+}
+
+impl VcpuThrottleThreadHandle {
+    /// The timeslice for a throttling cycle (vCPU pause & resume).
+    const TIMESLICE_MS: u64 = 500; // QEMU uses here 5000ms
+
+    /// This should not be named "vcpu*" as libvirt fails when
+    /// iterating the vCPU threads then. Fix this first in libvirt!
+    const THREAD_NAME: &'static str = "throttle-vcpu";
+
+    /// State change acknowledge by thread.
+    const CHANGE_ACK: StateChangeAckT = true;
+    /// State change not yet acknowledge by thread.
+    const CHANGE_NACK: StateChangeAckT = false;
+
+    /// Waits for the thread to ACK the state.
+    fn wait_for_thread_ack(&self, guard: MutexGuard<SynchronizedStateInner>) {
+        debug!("Waiting for thread to ACK waiting state");
+        let _guard = self
+            .shared_state
+            .1
+            .wait_while(guard, |(ack, _)| *ack == Self::CHANGE_ACK);
```

Comment:
I'm wodnering why this isn't `*ack == Self::CHANGE_NACK`. This is `wait_while` not `wait_until`. I have to double check this and add a comment here!